Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm

```diff
@@ -37,121 +37,205 @@
+    nw_listener_start(listener.get());
+
 }
 
 void NetworkTransportSession::sendDatagram(std::span<const uint8_t> data, CompletionHandler<void()>&& completionHandler)
 {
-    nw_connection_group_send_message(m_connectionGroup.get(), makeDispatchData(Vector(data)).get(), m_endpoint.get(), NW_CONNECTION_DEFAULT_MESSAGE_CONTEXT, makeBlockPtr([completionHandler = WTFMove(completionHandler)] (nw_error_t error) mutable {
+    nw_connection_send(m_connections.first().get(), makeDispatchData(Vector(data)).get(), NW_CONNECTION_DEFAULT_MESSAGE_CONTEXT, true, makeBlockPtr([completionHandler = WTFMove(completionHandler)] (nw_error_t error) mutable {
+        ASSERT(!error);
         // FIXME: Pass any error through to JS.
         completionHandler();
     }).get());
 }
 
 void NetworkTransportSession::createBidirectionalStream(CompletionHandler<void(std::optional<WebTransportStreamIdentifier>)>&& completionHandler)
 {
-    RetainPtr connection = adoptNS(nw_connection_group_extract_connection(m_connectionGroup.get(), m_endpoint.get(), nullptr));
+    createStream(NetworkTransportStreamType::Bidirectional, WTFMove(completionHandler));
+}
+
+void NetworkTransportSession::createOutgoingUnidirectionalStream(CompletionHandler<void(std::optional<WebTransportStreamIdentifier>)>&& completionHandler)
+{
+    createStream(NetworkTransportStreamType::OutgoingUnidirectional, WTFMove(completionHandler));
+}
+
+void NetworkTransportSession::setupListenerConnectionHandler()
+{
+    nw_listener_set_new_connection_handler(m_listener.get(), makeBlockPtr([weakThis = WeakPtr { *this }] (nw_connection_t inboundConnection) mutable {
+        ASSERT_UNUSED(weakThis, weakThis);
+        // FIXME: Implement incoming connection handler
+    }).get());
+}
+
+RetainPtr<nw_parameters_t> NetworkTransportSession::createParameters(NetworkTransportStreamType streamType)
+{
+    ASSERT(streamType != NetworkTransportStreamType::IncomingUnidirectional);
+    RetainPtr<nw_parameters_t> parameters = adoptNS(nw_parameters_copy(nw_connection_copy_parameters(m_connections.first().get())));
+    if (!parameters) {
+        ASSERT_NOT_REACHED();
+        return nil;
+    }
+    nw_protocol_stack_t default_stack = nw_parameters_copy_default_protocol_stack(parameters.get());
+    nw_protocol_stack_iterate_application_protocols(default_stack, [&streamType](nw_protocol_options_t options) {
+        if (nw_protocol_options_is_webtransport(options)) {
+            nw_webtransport_options_set_is_unidirectional(options, streamType == NetworkTransportStreamType::OutgoingUnidirectional);
+            nw_webtransport_options_set_is_datagram(options, false);
+        }
+    });
+    return parameters;
+}
+
+void NetworkTransportSession::receiveDatagramLoop()
+{
+    nw_connection_receive(m_connections.first().get(), 1, std::numeric_limits<uint32_t>::max(), makeBlockPtr([weakThis = WeakPtr { *this }] (dispatch_data_t content, nw_content_context_t, bool withFin, nw_error_t error) {
```

Comment:
We didn't check that m_connections has more than 0 connections in it, so we could be going out of bounds.  If the first connection is special, it should have its own member variable, like m_datagramConnection or something like that.